### PR TITLE
Havahol opencl2cuda

### PR DESCRIPTION
**Pull request by havahol**
Finally, done with the porting to CUDA, and two times messy-merge with all the refactoring going on in master.

Please check that you can run all tests:
```
cd gpu_ocean/tests
python basic_pycuda_tests.py 0
python oneLayer_2D_tests.py 0
python particle_filter_tests.py 0
python random_numbers.py 0
```

And that the following the notebook `gpu_ocean/prototypes/OceanModel_BasicParticleFilter.ipynb` runs.

### CODE REVIEW

File: gpu_ocean/SWESimulators/CTCS.py [hunk 1]
```diff
@@ -194,9 +194,12 @@ def fromfilename(cls, gpu_ctx, filename, cont_write_netcdf=True):
         timeIntegrator = sim_reader.get("time_integrator")
         y_zero_reference_cell = sim_reader.get("y_zero_reference_cell")
 
-        wind_stress_type = sim_reader.get("wind_stress_type")
-        wind = Common.WindStressParams(type=wind_stress_type)
-
+        try:
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/CTCS.py]: Shouldn't the sim_writer always write out the wind_stress_type?

File: gpu_ocean/SWESimulators/FBL.py [hunk 2]
```diff
@@ -195,8 +195,11 @@ def fromfilename(cls, gpu_ctx, filename, cont_write_netcdf=True):
         timeIntegrator = sim_reader.get("time_integrator")
         y_zero_reference_cell = sim_reader.get("y_zero_reference_cell")
 
-        wind_stress_type = sim_reader.get("wind_stress_type")
-        wind = Common.WindStressParams(type=wind_stress_type)
+        try:
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/FBL.py]: Again, I think we should require that the writer also writes out the stress type

File: gpu_ocean/SWESimulators/GPUDrifterCollection.py [hunk 3]
```diff
@@ -43,48 +43,56 @@ def __init__(self, cl_ctx, numDrifters, \
...
         # TODO: Where should the cl_queue come from?
         # For sure, the drifter and the ocean simulator should use 
         # the same queue...
-        self.cl_queue = cl_queue
-        if self.cl_queue is None:
-            self.cl_queue = pyopencl.CommandQueue(self.cl_ctx)
-        
+        self.gpu_stream = gpu_stream
+        if self.gpu_stream is None:
+            self.gpu_stream = cuda.Stream()
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/GPUDrifterCollection.py]: This should instead be the default argument, stream=cuda.Stream(). Right now, if I want to pass no stream (none), then I get a stream nonetheless

File: gpu_ocean/SWESimulators/OceanStateNoise.py [hunk 4]
```diff
@@ -142,39 +152,37 @@ def fromsim(cls, sim, soar_q0=None, soar_L=None,
...
-        return self.seed.download(self.cl_queue)
+        return self.seed.download(self.gpu_stream)
     
     def resetSeed(self):
+        # Generate seed:
         self.floatMax = 2147483648.0
         self.host_seed = np.random.rand(self.seed_ny, self.seed_nx)*self.floatMax
-        self.host_seed.astype(np.float32, order='C')
-        self.seed.upload(self.cl_queue, self.host_seed)
+        self.host_seed = self.host_seed.astype(np.uint64, order='C')
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/OceanStateNoise.py]: Is the type change here intentional?

File: gpu_ocean/SWESimulators/WindForcingEnsemble.py [hunk 5]
```diff
@@ -42,16 +43,21 @@ class WindForcingEnsemble(BaseOceanStateEnsemble.BaseOceanStateEnsemble):
...
-            wind = Common.WindStressParams(type=50, 
-                                           wind_speed=self.windSpeed,
-                                           wind_direction=self.directions[i])
-
             
-            self.particles[i] = CDKLM16.CDKLM16(self.cl_ctx, \
+            wX = [self.windX[i]*np.ones((2,2), dtype=np.float32)]
+            wY = [self.windY[i]*np.ones((2,2), dtype=np.float32)]
+            
+            wind = WindStress.WindStress(self.windT, wX, wY)
```
Comment (babrodtk) [inline gpu_ocean/SWESimulators/WindForcingEnsemble.py]: Wow, cool!

File: gpu_ocean/SWESimulators/CTCS.py (hunk 1 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/CTCS.py]: SimWriter and SimReader are not tested in this PR, as it needs non-trivial refactoring for the new wind stresses. Should we write all the wind fields to the NetCDF? Should we just refer to the original NetCDF, or make a new NetCDF if necessary and refer to that one?

I suggest to fix this issue in another PR. Added card in Trello: https://trello.com/c/fzDQPkwd

File: gpu_ocean/SWESimulators/FBL.py (hunk 2 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/FBL.py]: See comment above

File: gpu_ocean/SWESimulators/GPUDrifterCollection.py (hunk 3 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/GPUDrifterCollection.py]: Good idea. That would be cleaner. Sadly, it gives me an exception:
```File "../SWESimulators/GPUDrifterCollection.py", line 37, in GPUDrifterCollection
    gpu_stream=cuda.Stream(), \
pycuda._driver.LogicError: explicit_context_dependent failed: invalid device context - no currently active context?
```
I don't think the current implementation is a problem, though. Without a stream, you obviously can't do anything with the class until you attach the drifter to a Simulator. The gpu_stream in the Simulator object will then replace the stream created in line 54. The question is: How bad is it to create streams that are overwritten? The overwritten streams should go out of scope and be cleaned up by the garbage collector...

File: gpu_ocean/SWESimulators/OceanStateNoise.py (hunk 4 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/OceanStateNoise.py]: Yepp. The float32 here was actually a bug.

File: gpu_ocean/SWESimulators/WindForcingEnsemble.py (hunk 5 above)
Comment (havahol) [reply gpu_ocean/SWESimulators/WindForcingEnsemble.py]: Not cool, I would say :P The better solution would be to more flexible in the WindStress constructor. If X and Y was given by a list with a single float only (and not a list of np.arrays), the WindStress should create the 2x2 field with the constant value for you.

Comment (havahol) [reply gpu_ocean/SWESimulators/WindForcingEnsemble.py]: In a related note, I believe there's a bug in the wind stress... See: https://trello.com/c/jKzWJkIW

**Comment by havahol** [comment]
@babrodtk I've commented on all your comments. Let me know if you still think I should make changes

File: gpu_ocean/SWESimulators/CTCS.py (hunk 1 above)
Comment (setmar) [reply gpu_ocean/SWESimulators/CTCS.py]: I think it makes most sense to refer to the forcing input file(s) in global attributes. If an output file is used to hotstart a simulation, the forcing for already simulated time-steps is not very useful anyway.

**Comment by havahol** [comment]
Python 3 compatibility:  Won't fix in this PR. 

The notebook errors are known, and they are the reason why I specified some notebooks to check.

ParticleFilterWithChangingModel.ipynb:
Won't fix now. This notebooks depends heavily on how the wind stress acts on the velocity field. See https://trello.com/c/jKzWJkIW

OceanModel_SEnKF.ipynb:
There was a comment on the line above the one that failed explaining this fault. Won't fix.

ImplicitEqualWeightsParticleFilter.ipynb:
Prototype notebook. Won't fix.

**Comment by havahol** [comment]
Additional comment to @setmar: All the comments was also valid for the OpenCL version, and this PR therefore breaks no functionality, but it fixes quite a lot!

**Comment by setmar** [comment]
I suggest inserting a comment on top of notebooks with known errors.

**Comment by havahol** [comment]
Even though they are in prototypes?

**Comment by setmar** [comment]
I think it would be nice to know where known errors are, even in prototype notebooks. It makes them easier to use for others, and faster to pick up again for authors.

**Comment by setmar** [comment]
But no big deal. Just a thought.

**Comment by havahol** [comment]
Known errors come and go... I think my prototyping notebooks says a lot in the upper cells that have not been re-read and updated according to the prototyping results...

**Comment by babrodtk** [comment]
I'm happy with this, merging

**Pull request merged**